### PR TITLE
RFR - DRF Admin deletes comments

### PR DESCRIPTION
**Pull request by davidrf**

### CODE REVIEW

File: spec/features/user_destroys_review_spec.rb [hunk 1]
```diff
@@ -1,35 +1,35 @@
...
+    sign_in_as(user)
+    click_link beer.name
+    click_link "Delete Review"
+
+    expect(page).to have_content("Your review has been deleted!")
+  end
+
+  scenario "delete review without signing in" do
+    beer = FactoryGirl.create(:beer)
+    review = FactoryGirl.create(:review, beer: beer)
```
Comment (davidrf) [inline spec/features/user_destroys_review_spec.rb:28]: Used in line 33...

File: spec/features/user_edits_review_spec.rb [hunk 2]
```diff
@@ -1,53 +1,53 @@
...
+    click_link "Edit Review"
+    fill_in "Description", with: ""
+    click_button "Update Review"
+
+    expect(page).to have_content("Description can't be blank")
+  end
+
+  scenario "cannot review without signing in" do
+    beer = FactoryGirl.create(:beer)
+    review = FactoryGirl.create(:review, beer: beer)
```
Comment (davidrf) [inline spec/features/user_edits_review_spec.rb:46]: Used in line 51

File: app/models/review.rb [hunk 3]
```diff
@@ -11,7 +11,8 @@ class Review < ActiveRecord::Base
   validates :user, presence: true
   validates :beer, presence: true
 
-  def owner?(logged_in_user)
-    user == logged_in_user
+  def owner_or_admin?(logged_in_user)
+    return false unless logged_in_user
+    user == logged_in_user || "admin" == logged_in_user.role
```
Comment (vikram7) [inline app/models/review.rb]: This makes sense but is confusing to read. Can you refactor it so it's more clear?

File: app/views/beers/show.html.erb [hunk 4]
```diff
@@ -16,11 +16,11 @@ Description: <%= @beer.description %>
   <%= link_to "Edit Beer", edit_beer_path(@beer) %>
 <%-end-%>
 Reviews:
-<% @beer.reviews.each do |review| %>
+<% @beer.reviews.reverse.each do |review| %>
```
Comment (vikram7) [inline app/views/beers/show.html.erb]: is `.reverse` or sorting in activerecord/sql faster?

File: spec/features/admin_review_delete_spec.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  scenario "successfully deleted all reviews" do
+    admin = FactoryGirl.create(:user, role: "admin")
+    review = FactoryGirl.create(:review)
+
+    visit root_path
+    sign_in_as(admin)
+    click_link review.beer.name
+    click_link "Delete Review"
+
+    expect(page).to have_content("Your review has been deleted!")
```
Comment (vikram7) [inline spec/features/admin_review_delete_spec.rb:17]: and expect the page to not have the beer there, no?

File: spec/features/new_review_spec.rb [hunk 6]
```diff
@@ -1,52 +1,54 @@
...
+  # * I must be signed in
+  # * I must provide a rating and description
+  # * I must be presented with errors if I fill out the form incorrectly
+
+  scenario "successfully post review" do
+    beer = FactoryGirl.create(:beer)
+    user = beer.user
+
+    visit root_path
+    sign_in_as(user)
```
Comment (vikram7) [inline spec/features/new_review_spec.rb:18]: You want the same user to write a review on the beer they posted? Do you want to check if a different user can write a review just to make sure your logic work? You can write two separate tests for this.

File: spec/features/new_review_spec.rb [hunk 7]
```diff
@@ -1,52 +1,54 @@
...
+    expect(page).to have_content("Description can't be blank")
+  end
+
+  scenario "not signed in" do
+    beer = FactoryGirl.create(:beer)
+
+    visit root_path
+    click_link beer.name
+
+    expect(page).to_not have_link("New Review")
```
Comment (vikram7) [inline spec/features/new_review_spec.rb:70]: and expect to see "You must be signed in to do this"? Up to you if you want to add that.

File: spec/features/see_reviews_spec.rb [hunk 8]
```diff
@@ -1,23 +1,26 @@
...
+    reviews = FactoryGirl.create_list(:review, 2, beer: beer)
+    unrelated_review = FactoryGirl.create(:review)
+
+    visit root_path
+    click_link reviews.last.beer.name
+
+    new_review_position = page.body.index(reviews.last.description)
+    old_review_position = page.body.index(reviews.first.description)
+    expect(new_review_position).to be < old_review_position
+    expect(page).to_not have_content(unrelated_review.description)
```
Comment (vikram7) [inline spec/features/see_reviews_spec.rb:42]: Isn't this multiple tests?

`See reviews`
`See reviews in order I care about`
`Don't see reviews I shouldn't`

File: spec/features/user_destroys_review_spec.rb [hunk 9]
```diff
@@ -1,35 +1,36 @@
...
+
+    visit root_path
+    click_link beer.name
+
+    expect(page).to have_content(review.description)
+    expect(page).to_not have_link("Delete Review")
+  end
```
Comment (vikram7) [inline spec/features/user_destroys_review_spec.rb:35]: You also want to make sure a user who doesn't own the review can't delete it.

File: spec/features/user_edits_review_spec.rb [hunk 10]
```diff
@@ -1,53 +1,54 @@
...
+
+    visit root_path
+    click_link beer.name
+
+    expect(page).to have_content(review.description)
+    expect(page).to_not have_link("Edit Review")
+  end
+end
```
Comment (vikram7) [inline spec/features/user_edits_review_spec.rb:65]: Also check that different user can't edit.

File: spec/support/factory_girl.rb [hunk 11]
```diff
@@ -2,14 +2,14 @@
...
-    sequence(:email) {|n| "user#{n}@example.com" }
+    sequence(:email) { |n| "user#{n}@example.com" }
     password 'password'
     password_confirmation 'password'
     role 'user'
   end
 
   factory :beer do
-    name "Magic Hat"
+    sequence(:name) { |n| "Magic Hat #{n}" }
```
Comment (vikram7) [inline spec/support/factory_girl.rb:12]: Way better than Budweiser.

File: spec/support/factory_girl.rb [hunk 12]
```diff
@@ -2,14 +2,14 @@
...
     abv 5
     ibu 60
     description "Super hoppy goodness"
```
Comment (vikram7) [inline spec/support/factory_girl.rb:15]: Not all Magic Hat's are hoppy ;)

File: spec/features/admin_review_delete_spec.rb (hunk 5 above)
Comment (davidrf) [inline spec/features/admin_review_delete_spec.rb:17]: Added that expectation on line 18

File: spec/features/new_review_spec.rb (hunk 6 above)
Comment (davidrf) [inline spec/features/new_review_spec.rb:18]: Great suggestion, added scenario on line 31

File: spec/features/user_destroys_review_spec.rb (hunk 9 above)
Comment (davidrf) [inline spec/features/user_destroys_review_spec.rb:35]: scenario added in line 37

File: spec/features/user_edits_review_spec.rb [hunk 13]
```diff
@@ -1,53 +1,64 @@
...
+    visit root_path
+    click_link beer.name
+
+    expect(page).to have_content(review.description)
+    expect(page).to_not have_link("Edit Review")
+  end
+
+  scenario "user not review poster" do
+    beer = FactoryGirl.create(:beer)
+    another_user = FactoryGirl.create(:user)
```
Comment (davidrf) [inline spec/features/user_edits_review_spec.rb:57]: used in line 60

File: spec/features/new_review_spec.rb (hunk 7 above)
Comment (davidrf) [inline spec/features/new_review_spec.rb:70]: Right now there's no button for users not signed in to click, but maybe we should show it. Our group will discuss this and decide on it later.

File: app/models/beer.rb [hunk 14]
```diff
@@ -8,8 +8,7 @@ class Beer < ActiveRecord::Base
   validates :user, presence: true
   validates :description, length: { maximum: 5000 }
 
-  def owner_or_admin?(logged_in_user)
-    return false unless logged_in_user
-    user == logged_in_user || "admin" == logged_in_user.role
+  def amendable?(logged_in_user)
+    logged_in_user == user || logged_in_user.admin?
```
Comment (vikram7) [inline app/models/beer.rb:12]: nice

File: app/models/review.rb [hunk 15]
```diff
@@ -11,7 +11,7 @@ class Review < ActiveRecord::Base
   validates :user, presence: true
   validates :beer, presence: true
 
-  def owner?(logged_in_user)
-    user == logged_in_user
+  def amendable?(logged_in_user)
+    logged_in_user == user || logged_in_user.admin?
```
Comment (vikram7) [inline app/models/review.rb:15]: Ok you have this same method twice. Do you want to extract it out into a module somewhere? Not necessary right now, but put it on your list for refactoring. Once you get your tests running and working, you can refactor to your hearts' delights and run tests after each refactor.

File: spec/features/new_review_spec.rb [hunk 16]
```diff
@@ -1,52 +1,72 @@
...
+  As a user,
+  I want to post a review for a beer
+  so that I can influence the minds of others
+} do
+  # Acceptance Criteria
+  # * I must be signed in
+  # * I must provide a rating and description
+  # * I must be presented with errors if I fill out the form incorrectly
+
+  scenario "successfully post review" do
```
Comment (vikram7) [inline spec/features/new_review_spec.rb:13]: rename to `scenario "successfully post review on beer a user created"` to differentiate from the other one.

**Pull request merged**